design/debug-collector.md

```diff
@@ -0,0 +1,68 @@
+
+We would have to add this attribute to `DynamicHistory` and `ChunkedReportState` in ekotrace as well as to `FixedSliceVec` in fixed-slice-vec.
+
+## Symbol resolution
+
+For each tracer, the user will either pass in the memory location of a pointer to the tracer, or will pass in the name of the symbol of the pointer to the tracer. If the latter is used, then the user would also need to pass in an ELF file containing the symbol table of the program running on the target device. The debug-collector will automatically find the memory location of the given symbol by looping through the given symbol table.
+
+## Locking
+
+The debug-collector reads and writes to the ekotrace logs that the tracers running on the device have access to, so the debug-collector must ensure that it is not accessing logs at the same time as tracers. To do so, we can add an `in_use: Atomic<bool>` field to the DynamicHistory struct that will be set to true while the tracer is accessing the log. When the debug-collector pulls the logs, it will halt the CPU of the target and then check the value of `in_use`, aborting if true. If `in_use` is false, then the debug-collector can safely access the logs and then unhalt the CPU when finished. The use of atomic CPU instructions by the `Atomic<bool>` type should prevent race conditions.
+
+Since the debug-collector attempts to pull log data from multiple tracers at once, every iteration, it is possible that some of the tracers are locked. To deal with this, the debug-collector will simply skip the locked tracers. Depending on the logging frequency of tracers, this strategy could occasionally result in waiting multiple iterations to pull a given tracer's logs.
+
+# CLI design
+Tracer memory locations or symbols are inputted as positional arguments. If the argument starts with 0x, the CLI interprets the rest of the argument as a hex address of a tracer reference. Otherwise, the CLI interprets the argument as a symbol of a tracer reference, in which case the program binary must be included.
+
+Example:
+
+`ekotrace-debug-collector-cli --session-id 0 --word-size 32 --little-endian --elf ./target-elf --attach stm32 --interval 500 --output-file ./out.csv 0x20001000 TRACER_1 0x20000000 TRACER_3 TRACER_4`
+
+Flags:
+- `--session-id` `-s`: a session id to distinguish the trace data collected during this tracing run
+- `--word-size` `-w`: processor word size of the target system. If not given, detected from elf file if possible. Defaults to 32 with warning.
+- `--little-endian` `-le`; `--big-endian` `-be`: endianness of target system. If not given, detected from elf file if possible. Defaults to little with warning.
+- `--elf` `-e`: ELF file to use for symbol resolution and/or architecture detection
+- `--attach` `-a`: chip type passed to probe-rs
+- `--gdb-server` `-g`: address of gdb server attached to chip
+- `--interval` `-i`: interval between log report attempts
```

Comment:
Can we make flexible, something like [parse_duration](https://docs.rs/humantime/2.0.0/humantime/struct.Duration.html).